packages/devtools_app/lib/src/flutter/banner_messages.dart

```diff
@@ -68,62 +101,76 @@ class _BannerMessageContainerState extends State<BannerMessageContainer>
     final newController = Controllers.of(context)?.bannerMessages;
     if (newController == controller) return;
     controller = newController;
+  }
+
+  void push(BannerMessage message) {
+    if (controller.isMessageDismissed(message) ||
+        controller.isMessageVisible(message)) return;
+    // We push the banner message in a post frame callback because otherwise,we'd be
+    // trying to call setState while the parent widget `BannerMessages` is in the middle
+    // of `build`.
+    WidgetsBinding.instance.addPostFrameCallback((_) {
+      if (!mounted) return;
+      setState(() {
+        controller.addMessage(message);
+      });
+    });
+  }
 
-    autoDispose(controller.onRefreshMessages.listen((_) => setState(() {})));
+  void remove(BannerMessage message, {bool dismiss = false}) {
+    // We push the banner message in a post frame callback because otherwise,we'd be
+    // trying to call setState while the parent widget `BannerMessages` is in the middle
+    // of `build`.
+    WidgetsBinding.instance.addPostFrameCallback((_) {
+      if (!mounted) return;
+      setState(() {
+        controller.removeMessage(message, dismiss: dismiss);
+      });
+    });
   }
 
   @override
   Widget build(BuildContext context) {
-    final messagesForScreen = widget.screen.messages(context);
-    if (messagesForScreen.isNotEmpty) {
-      final messagesToShow = <Widget>[];
-      for (Widget message in messagesForScreen) {
-        assert(message is UniqueMessage);
-        if (!controller.isMessageDismissed((message as UniqueMessage).id)) {
-          messagesToShow.add(message);
-        }
-      }
-      if (messagesToShow.isNotEmpty) {
-        return Column(
-          children: messagesToShow,
-        );
-      }
-    }
-    return const SizedBox();
+    return _InheritedBannerMessages(
+      data: this,
+      child: Column(
+        children: [
+          ...controller?.messagesForScreen(widget.screen.type) ?? [],
```

Comment:
refactoring this to
```dart
Column(
  children: [
    Column(
      children: ValueListenableBuilder(controller?.messagesForScreen(widget.screen.type),
    ),
    Expanded(...),
)
```
will fix your issue with having to do a post frame callback and will make things work correctly if someone else uses the banner controller.

Reply:
doing this still results in the `setState() or markNeedsBuild() called during build.` error.

Reply:
darn. I guess the right mental model is you can't change any ValueListenable objects during build either.